src/sentry/utils/sentryappwebhookrequests.py

```diff
@@ -0,0 +1,128 @@
+        """
+        Add the item to the buffer key specified, using the given pipeline.
+        This does not execute the pipeline's commands.
+        """
+
+        pipeline.lpush(buffer_key, json.dumps(item))
+        pipeline.ltrim(buffer_key, 0, BUFFER_SIZE - 1)
+
+    def add_request(self, response_code, org_id, event, url):
+        if event not in VALID_EVENTS:
+            return
+
+        request_key = self._get_redis_key(event)
+
+        time = timezone.now()
+        request_data = {
+            "date": six.binary_type(time),
+            "response_code": response_code,
+            "webhook_url": url,
+        }
+
+        # Don't store the org id for internal apps because it will always be the org that owns the app anyway
+        if not self.sentry_app.is_internal:
+            request_data["organization_id"] = org_id
+
+        pipe = self.client.pipeline()
+
+        self._add_to_buffer_pipeline(request_key, request_data, pipe)
+
+        # If it's an error add it to the error buffer
+        if 400 <= response_code <= 599:
+            error_key = self._get_redis_key(event, error=True)
+            self._add_to_buffer_pipeline(error_key, request_data, pipe)
+
+        pipe.execute()
+
+    def _get_all_from_buffer(self, buffer_key, pipeline=None):
+        """
+        Get the list at the buffer key, using the given pipeline if available.
+        If a pipeline is provided, this does not return a value as the pipeline must still be executed.
+        """
+
+        if pipeline is not None:
+            pipeline.lrange(buffer_key, 0, BUFFER_SIZE - 1)
+        else:
+            return self.client.lrange(buffer_key, 0, BUFFER_SIZE - 1)
+
+    def _get_requests(self, event=None, error=False):
+        # If no event is specified, return the latest requests/errors for all event types
+        if event is None:
+            pipe = self.client.pipeline()
+
+            all_requests = []
+            for evt in VALID_EVENTS:
+                self._get_all_from_buffer(self._get_redis_key(evt, error=error), pipeline=pipe)
+
+            values = pipe.execute()
+
+            for idx, evt in enumerate(VALID_EVENTS):
+                event_requests = [
+                    self._convert_redis_request(request, evt) for request in values[idx]
+                ]
+                all_requests.extend(event_requests)
+
+            all_requests.sort(key=lambda x: parse_date(x.get("date")), reverse=True)
+            return all_requests[0:BUFFER_SIZE]
+
+        else:
+            return [
+                self._convert_redis_request(request, event)
+                for request in self._get_all_from_buffer(self._get_redis_key(event, error=error))
+            ]
+
+    def get_errors(self, event=None):
+        return self._get_requests(event=event, error=True)
+
+    def get_requests(self, event=None):
+        return self._get_requests(event=event, error=False)
+
+    def _get_redis_key(self, event, error=False):
```

Comment:
If we're targeting redis-cluster too, we need to make sure the key names are explicitly partitioned by `sentry_app_id`.

https://redis.io/topics/cluster-spec#keys-hash-tags

I'm not sure if there's better documentation somewhere over this, but we use it already elsewhere, so should be easy to find.

I could be wrong here, but I think ideally you'd want all redis data for a single sentry_app_id to be colocated on the same Redis shard. But maybe you don't need this. I'm making an assumption and sharing information so you can make an informed decision.